File: src/java/Integration.java
/*
 * Event Platform Client (EPC)
 *
 *     Designed for use with MediaWiki browser clients producing events to
 *     the EventGate intake service.
 *
 * LICENSE NOTICE
 *     Copyright 2019 Wikimedia Foundation
 *
 *     Redistribution and use in source and binary forms, with or without
 *     modification, are permitted provided that the following conditions are
 *     met:
 *
 *     1. Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *
 *     2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *
 *     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 *     IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 *     THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *     PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 *     CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 *     EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *     PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 *     PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 *     LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 *     NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *     SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * AUTHORS
 *     Jason Linehan <jlinehan@wikimedia.org>
 *     Mikhail Popov <mpopov@wikimedia.org>
 */

import org.json.*;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.TimeZone;
import java.util.UUID;

/******************************************************************************
 * Collect functions that will be replaced or mapped to other platform-specific
 * functions.
 ******************************************************************************/
class Integration {
    public static String get_device_id() {
        return "app install ID";
    }

    public static String get_stream_config() {
        return "{\"edit\":{\"scope\":\"session\",\"sample\":{\"rate\":0.06},\"destination\":\"https://pai-test.wmflabs.org/log\",\"$schema\":\"/edit/attempt-step/1.0.0\"}}";
    }

    public static String generate_id() {
        Random r = new Random();
        String id = "";
        for (int i = 0; i < 5; i++) {
            id += String.format("%04x", r.nextInt(65535));
        }
        return id;
    }

    public static String generate_uuid_v4() {
        return UUID.randomUUID().toString();
    }

    public static String generate_iso_8601_timestamp() {
        /**
         * Replace w/ DateUtil.getIso8601LocalDateFormat from org.wikipedia.util
         */
        TimeZone tz = TimeZone.getTimeZone("UTC");
        DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm'Z'");
        df.setTimeZone(tz);
        String iso8601 = df.format(new Date());
        return iso8601;
    }

    public static boolean client_cannot_be_tracked() {
        return false;
    }

    // HTTP POST request
    public static void http_post(String url, String body) throws Exception {
        URL obj = new URL(url);
        HttpURLConnection con = (HttpURLConnection) obj.openConnection();

        // add request header
        con.setRequestMethod("POST");
        con.setRequestProperty("User-Agent", "Wikimedia Product Analytics Infrastructure/Event Platform Client/Java 0.1");
        con.setRequestProperty("Accept-Language", "en-US,en");

        // Send post request
        con.setDoOutput(true);
        DataOutputStream wr = new DataOutputStream(con.getOutputStream());
        wr.writeBytes(body);
        wr.flush();
        wr.close();

        int responseCode = con.getResponseCode();
        System.out.println("\nSending 'POST' request to URL : " + url);
        System.out.println("Post body: " + body);
        System.out.println("Response Code : " + responseCode);

        BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
        String inputLine;
        StringBuffer response = new StringBuffer();

        while ((inputLine = in.readLine()) != null) {
            response.append(inputLine);
        }
        in.close();

        // print result
        System.out.println(response.toString());
    }

    public static <T> void set_persistent(String key, T value) {
        /* Do nothing */
        return;
    }

    public static <T> T get_persistent(String key) {
        /* Do nothing */
        return null;
    }

    public static <T> void del_persistent(String key) {
        /* Do nothing */
        return;
    }

    public static Map<String, Object> jsonToMap(JSONObject json) throws JSONException {
        Map<String, Object> retMap = new HashMap<String, Object>();

        if (json != JSONObject.NULL) {
            retMap = toMap(json);
        }

        return retMap;
    }

    public static Map<String, Object> toMap(JSONObject object) throws JSONException {
        Map<String, Object> map = new HashMap<String, Object>();

        Iterator<String> keysItr = object.keys();

        while (keysItr.hasNext()) {
            String key = keysItr.next();
            Object value = object.get(key);

            if (value instanceof JSONArray) {
                value = toList((JSONArray) value);
            } else if (value instanceof JSONObject) {
                value = toMap((JSONObject) value);
            }
            map.put(key, value);
        }
        return map;
    }

    public static List<Object> toList(JSONArray array) throws JSONException {
        List<Object> list = new ArrayList<Object>();

        for (int i = 0; i < array.length(); i++) {
            Object value = array.get(i);

            if (value instanceof JSONArray) {
                value = toList((JSONArray) value);
            } else if (value instanceof JSONObject) {
                value = toMap((JSONObject) value);
            }
            list.add(value);
        }
        return list;
    }
}

File: src/swift/EPC/Sources/EPC/Integration.swift
/*
 * Event Platform Client (EPC)
 *
 *     Designed for use with Wikipedia iOS application producing events to
 *     the EventGate intake service.
 *
 * LICENSE NOTICE
 *     Copyright 2019 Wikimedia Foundation
 *
 *     Redistribution and use in source and binary forms, with or without
 *     modification, are permitted provided that the following conditions are
 *     met:
 *
 *     1. Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *
 *     2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *
 *     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 *     IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 *     THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *     PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 *     CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 *     EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *     PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 *     PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 *     LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 *     NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *     SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * AUTHORS
 *     Mikhail Popov <mpopov@wikimedia.org>
 *     Jason Linehan <jlinehan@wikimedia.org>
 */

import Foundation

/**
 * Collect functions that will be replaced or mapped to other platform-specific functions.
 */
@available(iOS 10, OSX 10.12, *)
class Integration {

    // MARK: - Properties

    public static let shared = Integration() // singleton

    fileprivate let session: URLSession
    private let iso8601_formatter = ISO8601DateFormatter()

    /**
     * Store events until the library is finished initializing.
     *
     * The EPC library makes an HTTP request to a remote stream configuration service for information
     * about how to evaluate incoming event data. Until this initialization is complete, we store any incoming
     * events in this buffer.
     */
    private var input_buffer = [(stream: String, data: [String: Any])]()

    // MARK: - Methods

    private init(URLSession: URLSession = .shared) {
        self.session = URLSession
    }

    /**
     * Generates a new identifier using the same algorithm as EPC libraries for web and Android.
     */
    public func generate_id() -> String {
        var id: String = ""
        for _ in 1...8 {
            id += String(format: "%04x", arc4random_uniform(65535))
        }
        return id
    }

    /**
     * Append a `(stream, data)` pair to a queue (before stream configuration is available).
     */
    public func input_buffer_enqueue(_ item: (stream: String, data: [String: Any])) -> Void {
        input_buffer.append(item)
    }

    /**
     * Pluck oldest event from queue FIFO style (usually after stream configuration is available)
     */
    public func input_buffer_dequeue() -> (stream: String, data: [String: Any])? {
        if input_buffer.count == 0 {
            return nil
        } else {
            return input_buffer.remove(at: 0)
        }
    }

    public func client_cannot_be_tracked() -> Bool {
        return false
    }

    public func get_user_agent() -> String {
        // potentially dynamically fetch version info from a plist or other resource?
        return "Wikimedia Product Analytics Infrastructure/Event Platform Client/iOS 0.1"
    }

    /**
     * Fetches the app install ID.
     */
    public func device_id() -> String {
        let did: Any? = get_persistent("did") // like sid & pid
        if did != nil {
            return did as! String
        } else {
            return "<app install ID>"
        }
    }

    /**
     * Attempt to download the stream configuration from remote endpoint located and then relay it to the
     * stream manager.
     *
     * - Parameter url: stream configuration service public endpoint
     * - Parameter delegate: who to give the downloaded stream configuration to
     */
    public func load_stream_config(url: String, delegate: StreamManager) -> Void {
        var request = URLRequest(url: URL(string: url)!)
        request.httpMethod = "GET"
        request.setValue(get_user_agent(), forHTTPHeaderField: "User-Agent")
        let task = self.session.dataTask(with: request) {
            data, response, error in
            if let response = response as? HTTPURLResponse, let data = data {
                if response.statusCode == 200 || response.statusCode == 304 {
                    let from_json = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: [String: Any]]
                    if from_json != nil {
                        delegate.set_stream_config(from_json!)
                    } else {
                        print("Problem processing stream config from response")
                    }
                } else {
                    print("Bad response from endpoint: \(response.statusCode)")
                }
            }

        }
        task.resume()
    }

    public func generate_uuid_v4() -> String {
        // Initializes & returns a new UUID with RFC 4122 version 4 random bytes
        return UUID().uuidString
    }
    public func generate_iso_8601_timestamp() -> String {
        return iso8601_formatter.string(from: Date())
    }

    private func http_req(url: String, body: String, method: String) {
        var request = URLRequest(url: URL(string: url)!)
        let data = body.data(using: String.Encoding.utf8)
        request.httpMethod = method
        request.httpBody = data
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(get_user_agent(), forHTTPHeaderField: "User-Agent")
        let task = self.session.dataTask(with: request) { _, response, _ in
            if let httpResponse = response as? HTTPURLResponse {
                print("Response from \(url): \(httpResponse.statusCode)")
            }
        }
        task.resume()
    }

    /**
     * HTTP POST to public EventGate endpoint located at `url`
     */
    public func http_get(_ url: String, _ body: String) -> Void {
        http_req(url: url, body: body, method: "GET")
    }

    /**
     * HTTP GET to public EventGate endpoint located at `url`
     */
    public func http_post(_ url: String, _ body: String) -> Void {
        http_req(url: url, body: body, method: "POST")
    }

    /**
     * Fetch the value stored under key from persistent storage
     */
    public func get_persistent(_ key: String) -> Any? {
        // Integration with existent persistent storage data retrieval methods
        return nil
    }
    /**
     * Save the value stored under key to persistent storage
     */
    public func set_persistent(_ key: String, _ value: Any) -> Void {
        // Integration with existent persistent storage data retrieval methods
    }
    /**
     * Delete the value stored under key in persistent storage
     */
    public func del_persistent(_ key: String) -> Void {
        // Integration with existent persistent storage data retrieval methods
    }
}
